File: tree/benchmarks/higgs.py
from pandas.api.types import CategoricalDtype
from sklearn.metrics import accuracy_score, roc_auc_score
import argparse
import numpy as np
import pandas as pd
import json

parser = argparse.ArgumentParser()
parser.add_argument('--library', choices=['h2o', 'lightgbm', 'sklearn', 'xgboost', 'catboost'], required=True)
args = parser.parse_args()

# Load the data.
# path_train = 'data/higgs_500k_train.csv'
# path_test = 'data/higgs_500k_test.csv'
path_train = 'data/higgs_train.csv'
path_test = 'data/higgs_test.csv'
target_column_name = "signal"
dtype = {
	'signal': np.bool,
	'lepton_pt': np.float64,
	'lepton_eta': np.float64,
	'lepton_phi': np.float64,
	'missing_energy_magnitude': np.float64,
	'missing_energy_phi': np.float64,
	'jet_1_pt': np.float64,
	'jet_1_eta': np.float64,
	'jet_1_phi': np.float64,
	'jet_1_b_tag': np.float64,
	'jet_2_pt': np.float64,
	'jet_2_eta': np.float64,
	'jet_2_phi': np.float64,
	'jet_2_b_tag': np.float64,
	'jet_3_pt': np.float64,
	'jet_3_eta': np.float64,
	'jet_3_phi': np.float64,
	'jet_3_b_tag': np.float64,
	'jet_4_pt': np.float64,
	'jet_4_eta': np.float64,
	'jet_4_phi': np.float64,
	'jet_4_b_tag': np.float64,
	'm_jj': np.float64,
	'm_jjj': np.float64,
	'm_lv': np.float64,
	'm_jlv': np.float64,
	'm_bb': np.float64,
	'm_wbb': np.float64,
	'm_wwbb': np.float64,
}
data_train = pd.read_csv(path_train, dtype=dtype)
data_test = pd.read_csv(path_test, dtype=dtype)
features_train = data_train.loc[:, data_train.columns != target_column_name]
labels_train = data_train[target_column_name]
features_test = data_test.loc[:, data_test.columns != target_column_name]
labels_test = data_test[target_column_name]

# Train the model.
if args.library == 'h2o':
	import h2o
	from h2o.estimators import H2OGradientBoostingEstimator
	h2o.init(max_mem_size=20480000 * 1000)
	data_train = pd.concat([features_train, labels_train], axis=1)
	data_test = pd.concat([features_test, labels_test], axis=1)
	data_train = h2o.H2OFrame(python_obj=data_train)
	data_test = h2o.H2OFrame(python_obj=data_test)
	feature_column_names = [column for column in data_train.columns if column != target_column_name]
	model = H2OGradientBoostingEstimator(
		distribution="bernoulli",
		learn_rate=0.1,
		nbins=255,
		ntrees=100,
	)
	model.train(
		training_frame=data_train,
		x=feature_column_names,
		y=target_column_name,
	)
elif args.library == 'lightgbm':
	import lightgbm as lgb
	model = lgb.LGBMClassifier(
		force_row_wise=True,
		learning_rate=0.1,
		n_estimators=100,
		num_leaves=255,
	)
	model.fit(features_train, labels_train)
elif args.library == 'sklearn':
	from sklearn.experimental import enable_hist_gradient_boosting
	from sklearn.ensemble import HistGradientBoostingClassifier
	model = HistGradientBoostingClassifier(
		learning_rate=0.1,
		max_iter=100,
		max_leaf_nodes=255,
		validation_fraction=None,
	)
	model.fit(features_train, labels_train)
elif args.library == 'xgboost':
	import xgboost as xgb
	model = xgb.XGBClassifier(
		max_leaves=255,
		eta=0.1,
		grow_policy='lossguide',
		n_estimators=100,
		tree_method='hist',
	)
	model.fit(features_train, labels_train)
elif args.library == 'catboost':
	from catboost import CatBoostClassifier
	model = CatBoostClassifier(
		grow_policy='Lossguide',
		learning_rate=0.1,
		n_estimators=100,
		num_leaves=255,
		train_dir='data/catboost_info',
		verbose=False
	)
	model.fit(features_train, labels_train, silent=True)

# Make predictions on the test data.
if args.library == 'h2o':
	predictions_proba = model.predict(data_test).as_data_frame()['True']
else:
	predictions_proba = model.predict_proba(features_test)[:, 1]

# Compute metrics.
auc_roc = roc_auc_score(labels_test, predictions_proba)

# Compute memory usage.
f = open("/proc/self/status", "r")
for line in f.readlines():
	if line.startswith("VmHWM"):
		memory = line.split(":")[1].strip()

print(json.dumps({
	'auc_roc': auc_roc,
	'memory': memory,
}))

File: crates/tree/benchmarks/higgs.rs
use modelfox_table::prelude::*;
use modelfox_tree::Progress;
use modelfox_zip::{pzip, zip};
use ndarray::prelude::*;
use rayon::prelude::*;
use serde_json::json;
use std::{collections::BTreeMap, path::Path};

fn main() {
	// Load the data.
	let csv_file_path_train = Path::new("data/higgs_train.csv");
	let csv_file_path_test = Path::new("data/higgs_test.csv");
	let target_column_index = 0;
	let signal_variants = ["false", "true"].iter().map(ToString::to_string).collect();
	let options = modelfox_table::Options {
		column_types: Some(BTreeMap::from([
			(
				"signal".to_owned(),
				TableColumnType::Enum {
					variants: signal_variants,
				},
			),
			("lepton_pt".to_owned(), TableColumnType::Number),
			("lepton_eta".to_owned(), TableColumnType::Number),
			("lepton_phi".to_owned(), TableColumnType::Number),
			(
				"missing_energy_magnitude".to_owned(),
				TableColumnType::Number,
			),
			("missing_energy_phi".to_owned(), TableColumnType::Number),
			("jet_1_pt".to_owned(), TableColumnType::Number),
			("jet_1_eta".to_owned(), TableColumnType::Number),
			("jet_1_phi".to_owned(), TableColumnType::Number),
			("jet_1_b_tag".to_owned(), TableColumnType::Number),
			("jet_2_pt".to_owned(), TableColumnType::Number),
			("jet_2_eta".to_owned(), TableColumnType::Number),
			("jet_2_phi".to_owned(), TableColumnType::Number),
			("jet_2_b_tag".to_owned(), TableColumnType::Number),
			("jet_3_pt".to_owned(), TableColumnType::Number),
			("jet_3_eta".to_owned(), TableColumnType::Number),
			("jet_3_phi".to_owned(), TableColumnType::Number),
			("jet_3_b_tag".to_owned(), TableColumnType::Number),
			("jet_4_pt".to_owned(), TableColumnType::Number),
			("jet_4_eta".to_owned(), TableColumnType::Number),
			("jet_4_phi".to_owned(), TableColumnType::Number),
			("jet_4_b_tag".to_owned(), TableColumnType::Number),
			("m_jj".to_owned(), TableColumnType::Number),
			("m_jjj".to_owned(), TableColumnType::Number),
			("m_lv".to_owned(), TableColumnType::Number),
			("m_jlv".to_owned(), TableColumnType::Number),
			("m_bb".to_owned(), TableColumnType::Number),
			("m_wbb".to_owned(), TableColumnType::Number),
			("m_wwbb".to_owned(), TableColumnType::Number),
		])),
		..Default::default()
	};
	let mut features_train =
		Table::from_path(csv_file_path_train, options.clone(), &mut |_| {}).unwrap();
	let labels_train = features_train.columns_mut().remove(target_column_index);
	let mut features_test =
		Table::from_path(csv_file_path_test, options.clone(), &mut |_| {}).unwrap();
	let labels_test = features_test.columns_mut().remove(target_column_index);
	let labels_train = labels_train.as_enum().unwrap();
	let labels_test = labels_test.as_enum().unwrap();

	// Train the model.
	let train_options = modelfox_tree::TrainOptions {
		binned_features_layout: modelfox_tree::BinnedFeaturesLayout::RowMajor,
		learning_rate: 0.1,
		max_leaf_nodes: 255,
		max_rounds: 100,
		..Default::default()
	};
	let train_output = modelfox_tree::BinaryClassifier::train(
		features_train.view(),
		labels_train.view(),
		&train_options,
		Progress {
			kill_chip: &modelfox_kill_chip::KillChip::default(),
			handle_progress_event: &mut |_| {},
		},
	);

	// Make predictions on the test data.
	let features_test = features_test.to_rows();
	let chunk_size =
		(features_test.nrows() + rayon::current_num_threads() - 1) / rayon::current_num_threads();
	let mut probabilities = Array::zeros(features_test.nrows());
	pzip!(
		features_test.axis_chunks_iter(Axis(0), chunk_size),
		probabilities.axis_chunks_iter_mut(Axis(0), chunk_size),
	)
	.for_each(|(features_test_chunk, probabilities_chunk)| {
		train_output
			.model
			.predict(features_test_chunk, probabilities_chunk);
	});

	// Compute metrics.
	let input = zip!(probabilities.iter(), labels_test.iter())
		.map(|(probability, label)| (*probability, label.unwrap()))
		.collect();
	let auc_roc = modelfox_metrics::AucRoc::compute(input);

	let output = json!({
		"auc_roc": auc_roc,
	});
	println!("{}", output);
}
